executor/common_linux.h

```diff
@@ -1570,6 +1570,293 @@ static long syz_init_net_socket(volatile long domain, volatile long type, volati
+
+	struct hci_ev_cmd_complete evt_hdr;
+	evt_hdr.ncmd = 1;
+	evt_hdr.opcode = opcode;
+
+	uint8 type = HCI_EVENT_PKT;
+
+	iv[0].iov_base = &type;
+	iv[0].iov_len = sizeof(type);
+	iv[1].iov_base = &hdr;
+	iv[1].iov_len = sizeof(hdr);
+	iv[2].iov_base = &evt_hdr;
+	iv[2].iov_len = sizeof(evt_hdr);
+	iv[3].iov_base = data;
+	iv[3].iov_len = data_len;
+
+	return writev(fd, iv, sizeof(iv) / sizeof(struct iovec));
+}
+
+static void process_command_pkt(int fd, char* buf, size_t buf_size)
+{
+	struct hci_command_hdr* hdr = (struct hci_command_hdr*)buf;
+	if (buf_size < sizeof(struct hci_command_hdr) ||
+	    hdr->plen != buf_size - sizeof(struct hci_command_hdr)) {
+		fail("invalid size: %zx\n", buf_size);
+		return;
+	}
+
+	switch (hdr->opcode) {
+	case HCI_OP_READ_BD_ADDR: {
+		struct hci_rp_read_bd_addr rp;
+		rp.status = 0;
+		memset(&rp.bdaddr, 0xaa, 6);
+		hci_send_event_cmd_complete(fd, hdr->opcode, &rp, sizeof(rp));
+		return;
+	}
+	case HCI_OP_READ_BUFFER_SIZE: {
+		struct hci_rp_read_buffer_size rp;
+		rp.status = 0;
+		rp.acl_mtu = 1021;
+		rp.sco_mtu = 96;
+		rp.acl_max_pkt = 4;
+		rp.sco_max_pkt = 6;
+		hci_send_event_cmd_complete(fd, hdr->opcode, &rp, sizeof(rp));
+		return;
+	}
+	}
+
+	char status[0xf9] = {0};
+	hci_send_event_cmd_complete(fd, hdr->opcode, status, sizeof(status));
+}
+
+static void* event_thread(void* arg)
+{
+	int epoll_fd = *(int*)arg;
+
+	while (1) {
+		struct epoll_event event;
+		if (epoll_pwait(epoll_fd, &event, 1, -1, NULL) < 0)
+			break;
+
+		char buf[1024] = {0};
+		size_t buf_size = read(event.data.fd, buf, sizeof(buf));
+		if (buf_size > 0 && buf[0] == HCI_COMMAND_PKT)
+			process_command_pkt(event.data.fd, buf + 1, buf_size - 1);
+	}
+
+	return NULL;
+}
+
+#define ACL_HANDLE 256
+#define HCI_DEV 0
+
+static void initialize_vhci()
+{
+#if SYZ_EXECUTOR
+	if (!flag_vhci_injection)
+		return;
+#endif
+
+	int hci_socket = syz_init_net_socket(AF_BLUETOOTH, SOCK_RAW, BTPROTO_HCI);
+	if (hci_socket < 0)
+		fail("syz_init_net_socket failed");
+
+	vhci_fd = open("/dev/vhci", O_RDWR);
+	if (vhci_fd == -1)
+		fail("open /dev/vhci failed");
+
+	int epoll_fd = epoll_create1(0);
+	if (epoll_fd == -1)
+		fail("epoll_create1 failed");
+
+	struct epoll_event event = {0};
+	event.events = EPOLLIN;
+	event.data.fd = vhci_fd;
+	if (epoll_ctl(epoll_fd, EPOLL_CTL_ADD, vhci_fd, &event))
+		fail("epoll_ctl failed");
+
+	pthread_t th;
+	if (pthread_create(&th, NULL, event_thread, &epoll_fd))
+		fail("pthread_create failed");
+
+	// Register hci device
+	uint8 setup_cmd[2];
+	setup_cmd[0] = HCI_VENDOR_PKT;
+	setup_cmd[1] = 0;
+	if (write(vhci_fd, setup_cmd, sizeof(setup_cmd)) < 0)
+		fail("write failed");
+
+	// Bring hci device up
+	int res, tries = 16;
+	do {
+		res = ioctl(hci_socket, HCIDEVUP, HCI_DEV);
+		if (res != 0) {
+			if (errno == ENODEV)
+				continue;
+			else if (errno == EALREADY)
+				break;
+			else
+				fail("ioctl(HCIDEVUP) failed");
+		}
+	} while (res != 0 && tries--);
+
+	if (tries == 0)
```

Comment:
I don't think this is working. You use post decrement above, this means when the condition in the while gets 0, tries is already set to -1, right? So this line won't fail when we exited after 16 retries.
But it can fail if actually get HCIDEVUP returned successfully but tries is already set to 0.

Which makes me wonder if we need any of this retry logic at all? Do you see any retries and ENODEV?
We don't seem to do anything with HCI_DEV before this line, so I would assume that ENODEV is either returned permanently or not at all.
What I observe is that I always get EALREADY on the first attempt even on just booted VM.

Reply:
Ah, I should be as careful with my own code as when I'm auditing ;)

I removed the retry logic since the HCI_VENDOR_PKT write should already ensure that that the device is registered and brought up. The ioctl(HCIDEVUP) is to be really safe, but it's enough to do it once.